worlds/sm64ex/Rules.py

```diff
@@ -1,67 +1,96 @@
-from ..generic.Rules import add_rule
-from .Regions import connect_regions, sm64_level_to_paintings, sm64_paintings_to_level, sm64_level_to_secrets, sm64_entrances_to_level, sm64_level_to_entrances
+from typing import Callable, Union, Dict, Set
 
-def shuffle_dict_keys(world, obj: dict) -> dict:
-    keys = list(obj.keys())
-    values = list(obj.values())
+from BaseClasses import MultiWorld
+from ..generic.Rules import add_rule, set_rule
+from .Locations import location_table
+from .Regions import connect_regions, sm64_level_to_paintings, sm64_paintings_to_level, sm64_level_to_secrets, \
+sm64_level_to_secrets, sm64_entrances_to_level, sm64_level_to_entrances, Entrances
+from .Items import action_item_table
+
+def shuffle_dict_keys(world, dictionary: dict) -> dict:
+    keys = list(dictionary.keys())
+    values = list(dictionary.values())
     world.random.shuffle(keys)
-    return dict(zip(keys,values))
+    return dict(zip(keys, values))
 
-def fix_reg(entrance_ids, entrance, destination, swapdict, world):
-    if entrance_ids[entrance] == destination: # Unlucky :C
-        rand = world.random.choice(swapdict.keys())
-        entrance_ids[entrance], entrance_ids[swapdict[rand]] = rand, entrance_ids[entrance]
-        swapdict[rand] = entrance_ids[entrance]
-        swapdict.pop(entrance)
+def fix_reg(entrance_map: Dict[Entrances, str], entrance: Entrances, destination: str,
+            swapdict: Dict[str, Entrances], world, invalid_regions: Set[str] = {}):
```

Comment:
If you are introducing `invalid_regions`, no need to keep `destination` parameter; just use that instead. (Put it as second parameter tho, not at the end)